### ISSUE TITLE
[VO + Safari] Slider "increment" feature doesn't work well with aria-valuetext

### ISSUE THREAD

**Issue by phet-steele** [phetsims/john-travoltage#239]
The command "VO + Space" can be used on the leg or arm to perform their default action (as VO documentation puts it). In the case of the leg and arm, which are sliders, the default action is to increment the slider by 1.

For a basic idea of how this works, try the slider here: https://jsfiddle.net/1j1y3s8p/

Using VO + Space will increment the slider by one (no known way to decrement). Using the arrow keys of course move the slider in either direction. Using any combination of VO + Space and arrow keys behaves as expected...then there is our sim:

How the leg currently works with "VO + Space":

1. From sim start up, focus on the leg. You will be read its position.
2. Using VO + Space will work fine to increment the leg here....until,
3. Use an arrow key to move the leg.
4. Pressing VO + Space will no longer work. @jessegreenberg I believe step 3 leading to step 4 is the crux of the problem. If you press VO + Space, you will hear the current position with the word "increment" preceding the message BUT the leg no longer moves. Notice that more presses of VO + Space will repeat the same message with the same position value.
5. After you've had your fun with VO + Space, use an arrow key to move the leg. The leg will jump to a position that is based on how many times you pressed VO + Space. For example, if you started at position -3 and pressed VO + Space 4 times, pressing the right arrow key will jump the leg to position 2 (-3 + 4 + 1[right arrow key] = 2). The problem with this is that you've moved through a large amount of positions without picking up any electrons.

Bonus bug: after using VO + Space on the leg/arm any number of times, then tabbing to the arm/leg, will cease messages coming from the latter appendage after the first couple (or few?) messages.

Bonus bonus bug: Similar to the first bonus bug, the appendage you are currently incrementing can also have its messages cease. I do not know how or why, but it might be tied to pressing VO + Space multiple times at an appendages max position.

I'm inclined to say we make it so an appendage can be moved with VO + Space. However @emily-phet, it may not be necessary. It seems easy for users to learn that arrow keys are the better way to move an appendage. Even after running into these weird issues with VO + Space, the sim will quickly recover after a few tabs/arrow presses here and there. I'm also willing to bet that this is not a widely used key command. I would imagine this key command would have been brought up much earlier in interviews/design if it were widely used. @emily-phet up to you to have @jessegreenberg investigate why this is happening. My vote is to go ahead and support this feature. Why deny usability?

Seen on macOS 10.12.4. For phetsims/tasks/issues/831.

**Comment by emily-phet** [comment]
@jessegreenberg I agree that it would be good to support this, given the following:
- there is some way to decrement a slider value without using the arrow keys. 
If not, all users will have to utilize arrow keys anyway (since this sim really requires back and forth slider motion for productive use), so in that case I don't think it's worth more than a quick investigation to understand the issue better.

**Comment by jessegreenberg** [comment]
The problem is that VoiceOver + Safari handles slider events like this:

When using the arrow keys the slider receives an ```input``` event and then a ```change``` event.  When using the increment command, the slider receives only a ```change``` event.

This logic:
```js
    var keyboardEventHandled = false;
    this.addAccessibleInputListener( {
      input: function( event ) {
        self.rotateAppendage();
        keyboardEventHandled = true;
        self.dragging = true;
      },
      change: function( event ) {
        if ( !keyboardEventHandled ) {
          self.rotateAppendage();
        }
        self.dragging = true;
      },
```

Is preventing the change event from doing anything.  This block of code was necessary because different browsers emit ```input``` and ```change``` events differently.

**Comment by jobara** [comment]
@jessegreenberg try adding `keyboardEventHandled = false;` at the end of the change event. I think what might be happening is that because there is a mix of cases "input and change" and "change only" that the `keyboardEventHandled` flag is always left at `true` if interacted with the arrow keys.

**Comment by jessegreenberg** [comment]
Thanks @jobara that seems to work really well!  After the change, we still have @phet-steele's bonus and bonus bonus bugs in the original issue comment.

**Comment by jessegreenberg** [comment]
When using the "increment" feature, VoiceOver is reading the *previous* aria-valuetext. So when we are at position 1, it says we are at position 0.

**Comment by jessegreenberg** [comment]
My guess is that when the "change" event happens, VoiceOver immediately reads the aria-valuetext and then emits events for listeners?

**Comment by jessegreenberg** [comment]
Here is an updated JSFiddle example which lets us see the aria-valuetext:

https://jsfiddle.net/1j1y3s8p/2/

Using the increment feature, the value is correct, but it also stops reading any aria-valuetext after incrementing.  That much seems like a VoiceOver problem.

**Comment by jessegreenberg** [comment]
Although sometimes the aria-valuetext is correct when using the increment feature.

**Comment by jessegreenberg** [comment]
I can't get the incorrect aria-valuetext to be read at all in the JSFiddle example.

**Comment by jessegreenberg** [comment]
aria-controls attribute that is causing this bug.  Here is a JSFiddle with this problem.

https://jsfiddle.net/1j1y3s8p/6/

But when I remove the aria-controls attribute from the html element in the sim, the problem persists.  I have deleted caches and restarted voiceover.

**Comment by jessegreenberg** [comment]
Here is a JSFiddle with the exact html (without aria-controls).  It has the same bug:

https://jsfiddle.net/1j1y3s8p/8/

**Comment by jessegreenberg** [comment]
Actually, that had aria-controls.  This one doesn't

https://jsfiddle.net/1j1y3s8p/12/

It still has the bug.

I notice that when we increment, the label is read.  I wonder if we remove the label if the problem goes away?

**Comment by jessegreenberg** [comment]
Yes, if we remove the label the problem goes away.

**Comment by jessegreenberg** [comment]
Removing the label doesn't fix the problem in the sim.

**Comment by jessegreenberg** [comment]
Bug is still present intermittently in https://jsfiddle.net/1j1y3s8p/15/.

I don't think we should spend much more time on this bug, seems like a VoiceOver issue with the way the "increment" feature is handled.

**Comment by jessegreenberg** [comment]
The fix recommended by @jobara gets the increment feature to work, although we still have the following VoiceOver bugs (2 of which were listed in the original issue)

- After using the increment feature the ```aria-valuetext``` often isn't read when using the arrow keys.
- After using the increment feature, the ```aria-valuetext``` often isn't read after tabbing away from the appendage.
- When using the increment feature, the ```aria-valuetext``` read by VO is often one behind the current value.

Here is a very simple JSFiddle with all of these problems as well: https://jsfiddle.net/1j1y3s8p/15/

@emily-phet @terracoda @jobara the VO "increment" feature is now working as well as VoiceOver can support, and I think the remaining problems should be ignored for now, do you agree?

**Comment by emily-phet** [comment]
@jessegreenberg I agree. Let's not work on this further for now.

**Comment by jessegreenberg** [comment]
Sounds good @emily-phet.  Marking as ```wontfix``` and changing the issue title to reflect what is currently the problem.

**Comment by jessegreenberg** [comment]
https://github.com/phetsims/a11y-research/issues/36 will track submitting a bug report to Apple.

**Comment by jessegreenberg** [comment]
This can be closed since it is being tracked in a11y-research#36.
